File: termux_aichain/memory/buffer.py
"""
==============================================================================
termux-aichain Memory Engine: Conversation Buffer Memory
==============================================================================
Provides short-term windowed conversation history management.
Zero external heavy dependencies - Pure Python 3.10+ standard library.
"""

from __future__ import annotations
from typing import Any, Dict, List, Optional, Union
from termux_aichain.core.schema import Message, HumanMessage, AIMessage, SystemMessage

class ConversationBufferMemory:
    """Maintains a rolling window of recent conversation messages."""

    def __init__(self, k: int = 10, return_messages: bool = True, memory_key: str = "history"):
        self.k = k
        self.return_messages = return_messages
        self.memory_key = memory_key
        self.chat_history: List[Message] = []

    def save_context(self, inputs: Union[Dict[str, Any], str], outputs: Union[Dict[str, Any], str]) -> None:
        user_text = inputs if isinstance(inputs, str) else list(inputs.values())[0] if inputs else ""
        ai_text = outputs if isinstance(outputs, str) else list(outputs.values())[0] if outputs else ""

        self.chat_history.append(HumanMessage(content=str(user_text)))
        self.chat_history.append(AIMessage(content=str(ai_text)))

        # Truncate to maximum 2 * k messages (k turns)
        if len(self.chat_history) > self.k * 2:
            self.chat_history = self.chat_history[-(self.k * 2):]

    def load_memory_variables(self, inputs: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        if self.return_messages:
            return {self.memory_key: list(self.chat_history)}
        # String representation
        lines = []
        for m in self.chat_history:
            role = "Human" if m.role == "user" else "AI" if m.role == "assistant" else m.role.title()
            lines.append(f"{role}: {m.content}")
        return {self.memory_key: "\n".join(lines)}

    def clear(self) -> None:
        self.chat_history.clear()

File: js/esm/memory/buffer.js
/**
 * ==============================================================================
 * @termux-ai/chain Memory Engine: ConversationBufferMemory (TypeScript ESM)
 * ==============================================================================
 */
import { HumanMessage, AIMessage } from "../core/schema.js";
export class ConversationBufferMemory {
    k;
    returnMessages;
    memoryKey;
    chatHistory = [];
    constructor(options = {}) {
        this.k = options.k ?? 10;
        this.returnMessages = options.returnMessages ?? true;
        this.memoryKey = options.memoryKey ?? "history";
    }
    saveContext(inputs, outputs) {
        const userText = typeof inputs === "string" ? inputs : Object.values(inputs)[0] ?? "";
        const aiText = typeof outputs === "string" ? outputs : Object.values(outputs)[0] ?? "";
        this.chatHistory.push(new HumanMessage(String(userText)));
        this.chatHistory.push(new AIMessage(String(aiText)));
        if (this.chatHistory.length > this.k * 2) {
            this.chatHistory = this.chatHistory.slice(-(this.k * 2));
        }
    }
    loadMemoryVariables() {
        if (this.returnMessages) {
            return { [this.memoryKey]: [...this.chatHistory] };
        }
        const lines = this.chatHistory.map(m => {
            const role = m.role === "user" ? "Human" : m.role === "assistant" ? "AI" : m.role;
            return `${role}: ${m.content}`;
        });
        return { [this.memoryKey]: lines.join("\n") };
    }
    clear() {
        this.chatHistory = [];
    }
}
